prov/efa/src/rxr/rxr_env.c

```diff
@@ -93,7 +96,13 @@ void rxr_env_param_get(void)
 
 	fi_param_get_int(&rxr_prov, "tx_queue_size", &rxr_env.tx_queue_size);
 	fi_param_get_int(&rxr_prov, "enable_shm_transfer", &rxr_env.enable_shm_transfer);
-	fi_param_get_int(&rxr_prov, "use_device_rdma", &rxr_env.use_device_rdma);
+	ret = fi_param_get_int(&rxr_prov, "use_device_rdma", &rxr_env.use_device_rdma);
+	if (ret != -FI_ENODATA && rxr_env.use_device_rdma &&
+	    !(g_device_list[0].efa_attr.device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ)) {
+		FI_WARN(&rxr_prov, FI_LOG_FABRIC, "FI_EFA_USE_DEVICE_RDMA=1 was set by users, "
```

Comment:
I think we probably abort() here and `fprintf(stderr)` the error message, instead of return ERROR. 

Because reeturning error may cause application to use tcp silently.

Reply:
I thought about that. This function is actually called during libfabric registering all providers. So if users happen to have this env set on a platform that does not support rdma-read, the libfabric will abort early and user cannot use any provider. I can make the change if we are ok with that.

Reply:
That seems bad.  What upper layer doesn't block the tcp providers anyway?

Reply:
Intel MPI will fallback to use tcp;ofi_rxm, if EFA provider is not available.

The abort only happens on a system with EFA device and FI_EFA_USE_DEVICE_RDMA was set explicitly. So I think it should be good.

Reply:
Do we also abort when it's 0xefa2 but firmware does not support rdma-read ?

Reply:
> 0xefa2 but firmware does not support rdma-read 

In that case, this patch exit early in `rxr_env_initialize()`. So we will not reach here.

That should not happen any way.

Reply:
I am asking whether we abort here

```
	int rxr_env_initialize()
{
	uint32_t vendor_part_id = g_device_list[0].ibv_attr.vendor_part_id;

	if (vendor_part_id == 0xefa0 || vendor_part_id == 0xefa1) {
		rxr_env.use_device_rdma = 0;
	} else {
		if (!(g_device_list[0].efa_attr.device_caps & EFADV_DEVICE_ATTR_CAPS_RDMA_READ)) {
			FI_WARN(&rxr_prov, FI_LOG_FABRIC,
				"EFA device with vendor_part_id 0x%x has no rdma-read capability.\n", vendor_part_id);
			return -FI_EINVAL;
		}
		rxr_env.use_device_rdma = 1;
	}
```

Reply:
I think we should not abort(), but we should not return error either.

I think we should print warning and proceed with `use_device_rdma` to 0.

If FI_EFA_USE_DEVICE_RDMA was set to 1,  abort()  there (in `rxr_env_param_get`).

If it is not set, we do not use read.

Reply:
OK, I updated the PR accordingly.